src/services/refactors/inlineVariable.ts

```diff
@@ -0,0 +1,194 @@
+                    notApplicableReason: info.error
+                }]
+            }];
+        }
+
+        if (!refactor.isRefactorErrorInfo(info)) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [inlineVariableAction]
+            }];
+        }
+
+        return emptyArray;
+    },
+
+    getEditsForAction(context, actionName) {
+        Debug.assert(actionName === refactorName, "Unexpected refactor invoked");
+
+        const { file, program, startPosition } = context;
+
+        const info = getInliningInfo(file, startPosition, /*tryWithReferenceToken*/ true, program);
+        if (!info || refactor.isRefactorErrorInfo(info)) {
+            return undefined;
+        }
+
+        const { references, declaration, replacement } = info;
+        const edits = textChanges.ChangeTracker.with(context, tracker => {
+            for (const node of references) {
+                tracker.replaceNode(file, node, getReplacementExpression(node, replacement));
+            }
+            tracker.delete(file, declaration);
+        });
+
+        return { edits };
+    }
+});
+
+function getInliningInfo(file: SourceFile, startPosition: number, tryWithReferenceToken: boolean, program: Program): InliningInfo | RefactorErrorInfo | undefined {
+    const checker = program.getTypeChecker();
+    const token = getTokenAtPosition(file, startPosition);
+    const parent = token.parent;
+
+    // If the node is a variable declaration, make sure it's not in a catch clause or for-loop
+    // and that it has a value.
+    if (isInitializedVariable(parent) && isVariableDeclarationInVariableStatement(parent) && isIdentifier(parent.name)) {
+        // Don't inline the variable if it isn't declared exactly once.
+        if (parent.symbol.declarations?.length !== 1) {
+            return { error: getLocaleSpecificMessage(Diagnostics.Variables_that_share_a_name_with_a_type_or_namespace_in_the_same_scope_cannot_be_inlined) };
+        }
+
+        // Do not inline if the variable is exported.
+        if (findAncestor(parent, node => canHaveModifiers(node) && some(node.modifiers, isExportModifier))) {
+            return undefined;
+        }
+
+        // Find all references to the variable in the current file.
+        const references = getReferenceNodes(parent, checker, file);
+        return references && { references, declaration: parent, replacement: parent.initializer };
+    }
+
+    if (tryWithReferenceToken && isIdentifier(token)) {
+        // Try finding the declaration and nodes to replace via the reference token.
+        const definition = checker.resolveName(token.text, token, SymbolFlags.Value, /*excludeGlobals*/ false);
+        if (definition?.declarations?.length !== 1) {
+            return { error: getLocaleSpecificMessage(Diagnostics.Variables_that_share_a_name_with_a_type_or_namespace_in_the_same_scope_cannot_be_inlined) };

```

Comment:
I wonder if this is hit at all - you're telling the checker to only resolve values, but the error here is about types and namespaces. I think it ties into my feedback above.

Reply:
Are you referring to your feedback about not using `findAncestor`?

Reply:
No, what I mean is that you are resolving the name, but are only allowing values:

```ts
const definition = checker.resolveName(token.text, token, SymbolFlags.Value, /*excludeGlobals*/ false);
//                                                        ^^^^^^^^^^^^^^^^^
```

But the user-issued error message is strictly talking about "types" and "namespaces".

So maybe you can hit this case, but the error message doesn't make sense here (nor does it above)

Reply:
Could you help me come up with a more suitable error message? :)